docs/configuration.md
# Configuration

There are two ways to configure Flipt: using a configuration file or through environment variables.

## Configuration File

The default way that Flipt is configured is with the use of a configuration file `default.yml`.

This file is read when Flipt starts up and configures several important properties for the server.

You can edit any of these properties to your liking, and on restart Flipt will pick up the new changes.

!!! note
    These defaults are commented out in `default.yml` to give you an idea of what they are. To change them you'll first need to uncomment them.

These properties are as follows:

| Property | Description | Default |
|---|---|---|
| log.level | Level at which messages are logged (trace, debug, info, warn, error, fatal, panic) | info |
| log.file | File to log to instead of STDOUT | |
| ui.enabled | Enable UI and API docs | true |
| cors.enabled | Enable CORS support | false |
| cors.allowed_origins | Sets Access-Control-Allow-Origin header on server | "*" (all domains) |
| cache.memory.enabled | Enable in-memory caching | false |
| cache.memory.items | Number of items in-memory cache can hold | 500 |
| server.protocol | `http` or `https` | http |
| server.host | The host address on which to serve the Flipt application | 0.0.0.0 |
| server.http_port | The HTTP port on which to serve the Flipt REST API and UI | 8080 |
| server.https_port | The HTTPS port on which to serve the Flipt REST API and UI | 443 |
| server.grpc_port | The port on which to serve the Flipt GRPC server | 9000 |
| server.cert_file | Path to the certificate file (if protocol is set to `https`) | |
| server.cert_key | Path to the certificate key file (if protocol is set to `https`) | |
| db.url | URL to access Flipt database | file:/var/opt/flipt/flipt.db |
| db.migrations.path | Where the Flipt database migration files are kept | /etc/flipt/config/migrations |

## Using Environment Variables

All options in the configuration file can be overridden using environment variables using the syntax:

```shell
FLIPT_<SectionName>_<KeyName>
```

!!! tip
    Using environment variables to override defaults is especially helpful when running with Docker as described in the [Installation](installation.md) documentation.

Everything should be upper case, `.` should be replaced by `_`. For example, given these configuration settings:

```yaml
server:
  grpc_port: 9000

db:
  url: file:/var/opt/flipt/flipt.db
```

You can override them using:

```shell
export FLIPT_SERVER_GRPC_PORT=9001
export FLIPT_DB_URL="postgres://postgres@localhost:5432/flipt?sslmode=disable"
```

## Databases

Flipt supports both [SQLite](https://www.sqlite.org/index.html) and [Postgres](https://www.postgresql.org/) databases as of `v0.5.0`.

SQLite is enabled by default for simplicity, however you should use Postgres if you intend to run multiple copies of Flipt in a high availability configuration.

The database connection can be configured as follows:

### SQLite

```yaml
db:
  # file: informs flipt to use SQLite
  url: file:/var/opt/flipt/flipt.db
```

### Postgres

```yaml
db:
  url: postgres://postgres@localhost:5432/flipt?sslmode=disable
```

!!! note
    The Postgres database must exist and be up and running before Flipt will be able to connect to it.

## Migrations

From time to time the Flipt database must be updated with new schema. To accomplish this, Flipt includes a `migrate` command that will run any pending database migrations for you.

If Flipt is started and there are pending migrations, you will see the following error in the console:

```bash
migrations pending, please backup your database and run `flipt migrate`
```

If it is your first run of Flipt, all migrations will automatically be run before starting the Flipt server.

!!! warning
    You should backup your database before running `flipt migrate` to ensure that no data is lost if an error occurs during migration.

If running Flipt via Docker, you can run the migrations in a separate container before starting Flipt by running:

```bash
docker run -it markphelps/flipt:latest /bin/sh -c './flipt migrate'
```

## Caching

### In-Memory

In-memory caching is currently only available for flags. When enabled, in-memory caching has been shown to speed up the fetching of individual flags by 10x.

To enable caching set the following in your config:

```yaml
cache:
  memory:
    enabled: true
```

Work is planned to add caching support to rule evaluation soon.

!!! warning
    Enabling in-memory caching when running more that one instance of Flipt is not advised as it will lead to unpredictable results.

## Metrics

Flipt exposes [Prometheus](https://prometheus.io/) metrics at the `/metrics` HTTP endpoint. To see which metrics are currently supported, point your browser to `FLIPT_HOST/metrics` (ex: `localhost:8080/metrics`).

You should see a bunch of metrics being recorded such as:

```text
flipt_cache_hit_total{cache="memory",type="flag"} 1
flipt_cache_miss_total{cache="memory",type="flag"} 1
...
go_gc_duration_seconds{quantile="0"} 8.641e-06
go_gc_duration_seconds{quantile="0.25"} 2.499e-05
go_gc_duration_seconds{quantile="0.5"} 3.5359e-05
go_gc_duration_seconds{quantile="0.75"} 6.6594e-05
go_gc_duration_seconds{quantile="1"} 0.00026651
go_gc_duration_seconds_sum 0.000402094
go_gc_duration_seconds_count 5
...
```

## Authentication

There is currently no built in authentication, authorization or encryption as Flipt was designed to work inside your trusted architecture and not be exposed publicly.

If you do wish to expose the Flipt dashboard and REST API publicly using HTTP Basic Authentication, you can do so by using a reverse proxy. There is an [example](https://github.com/markphelps/flipt/tree/master/examples/auth) provided in the GitHub repository showing how this could work.


CHANGELOG.md
# Changelog

This format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added

* Ability to write logs to file instead of STDOUT: [https://github.com/markphelps/flipt/issues/141](https://github.com/markphelps/flipt/issues/141)

## [v0.9.0](https://github.com/markphelps/flipt/releases/tag/v0.9.0) - 2019-10-02

### Added

* Support evaluating flags without variants: [https://github.com/markphelps/flipt/issues/138](https://github.com/markphelps/flipt/issues/138)

### Changed

* Dropped support for Go 1.12

### Fixed

* Segments not matching on all constraints: [https://github.com/markphelps/flipt/issues/140](https://github.com/markphelps/flipt/issues/140)
* Modal content streching to fit entire screen

## [v0.8.0](https://github.com/markphelps/flipt/releases/tag/v0.8.0) - 2019-09-15

### Added

* HTTPS support

## [v0.7.1](https://github.com/markphelps/flipt/releases/tag/v0.7.1) - 2019-07-25

### Added

* Exposed errors metrics via Prometheus

### Changed

* Updated JS dev dependencies
* Updated grpc/protobuf versions
* Updated pq version
* Updated go-sqlite3 version

## [v0.7.0](https://github.com/markphelps/flipt/releases/tag/v0.7.0) - 2019-07-07

### Added

* CORS support with `cors` config options
* [Prometheus](https://prometheus.io/) metrics exposed at `/metrics`

## [v0.6.1](https://github.com/markphelps/flipt/releases/tag/v0.6.1) - 2019-06-14

### Fixed

* Missing migrations folders in release archive: [https://github.com/markphelps/flipt/issues/97](https://github.com/markphelps/flipt/issues/97)

## [v0.6.0](https://github.com/markphelps/flipt/releases/tag/v0.6.0) - 2019-06-10

### Added

* `migrate` subcommand to run database migrations
* 'Has Prefix' and 'Has Suffix' constraint operators

### Changed

* Variant keys are now only required to be unique per flag, not globally: [https://github.com/markphelps/flipt/issues/87](https://github.com/markphelps/flipt/issues/87)

### Removed

* `db.migrations.auto` in config. DB migrations must now be run explicitly with the `flipt migrate` command

## [v0.5.0](https://github.com/markphelps/flipt/releases/tag/v0.5.0) - 2019-05-27

### Added

* Beta support for Postgres! :tada:
* `/meta/info` endpoint for version/build info
* `/meta/config` endpoint for running configuration info

### Changed

* `cache.enabled` config becomes `cache.memory.enabled`
* `cache.size` config becomes `cache.memory.items`
* `db.path` config becomes `db.url`

### Removed

* `db.name` in config

## [v0.4.2](https://github.com/markphelps/flipt/releases/tag/v0.4.2) - 2019-05-12

### Fixed

* Segments with no constraints now match all requests by default: [https://github.com/markphelps/flipt/issues/60](https://github.com/markphelps/flipt/issues/60)
* Clear Debug Console response on error

## [v0.4.1](https://github.com/markphelps/flipt/releases/tag/v0.4.1) - 2019-05-11

### Added

* `/debug/pprof` [pprof](https://golang.org/pkg/net/http/pprof/) endpoint for profiling

### Fixed

* Issue in evaluation: [https://github.com/markphelps/flipt/issues/63](https://github.com/markphelps/flipt/issues/63)

## [v0.4.0](https://github.com/markphelps/flipt/releases/tag/v0.4.0) - 2019-04-06

### Added

* `ui` config section to allow disabling the ui:

    ```yaml
    ui:
      enabled: true
    ```

* `/health` HTTP healthcheck endpoint

### Fixed

* Issue where updating a Constraint or Variant via the UI would not show the update values until a refresh: [https://github.com/markphelps/flipt/issues/43](https://github.com/markphelps/flipt/issues/43)
* Potential IndexOutOfRange error if distribution percentage didn't add up to 100: [https://github.com/markphelps/flipt/issues/42](https://github.com/markphelps/flipt/issues/42)

## [v0.3.0](https://github.com/markphelps/flipt/releases/tag/v0.3.0) - 2019-03-03

### Changed

* Renamed generated proto package to `flipt` for use with external GRPC clients
* Updated docs and example to reference [GRPC go client](https://github.com/markphelps/flipt-grpc-go)

### Fixed

* Don't return error on graceful shutdown of HTTP server

## [v0.2.0](https://github.com/markphelps/flipt/releases/tag/v0.2.0) - 2019-02-24

### Added

* `server` config section to consolidate and rename `host`, `api.port` and `backend.port`:

    ```yaml
    server:
      host: 0.0.0.0
      http_port: 8080
      grpc_port: 9000
    ```

* Implemented flag caching! Preliminary testing shows about a 10x speedup for retrieving flags with caching enabled. See the docs for more info.

    ```yaml
    cache:
      enabled: true
    ```

### Deprecated

* `host`, `api.port` and `backend.port`. These values have been moved and renamed under the `server` section and will be removed in the 1.0 release.

## [v0.1.0](https://github.com/markphelps/flipt/releases/tag/v0.1.0) - 2019-02-19

### Added

* Moved proto/client code to proto directory and added MIT License

## [v0.0.0](https://github.com/markphelps/flipt/releases/tag/v0.0.0) - 2019-02-16

Initial Release!


config/config.go
package config

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"

	"github.com/pkg/errors"
	"github.com/spf13/viper"
)

type Config struct {
	Log      logConfig      `json:"log,omitempty"`
	UI       uiConfig       `json:"ui,omitempty"`
	Cors     corsConfig     `json:"cors,omitempty"`
	Cache    cacheConfig    `json:"cache,omitempty"`
	Server   serverConfig   `json:"server,omitempty"`
	Database databaseConfig `json:"database,omitempty"`
}

type logConfig struct {
	Level string `json:"level"`
	File  string `json:"file,omitempty"`
}

type uiConfig struct {
	Enabled bool `json:"enabled"`
}

type corsConfig struct {
	Enabled        bool     `json:"enabled"`
	AllowedOrigins []string `json:"allowedOrigins,omitempty"`
}

type memoryCacheConfig struct {
	Enabled bool `json:"enabled"`
	Items   int  `json:"items,omitempty"`
}

type cacheConfig struct {
	Memory memoryCacheConfig `json:"memory,omitempty"`
}

type Scheme uint

func (s Scheme) String() string {
	return schemeToString[s]
}

const (
	HTTP Scheme = iota
	HTTPS
)

var (
	schemeToString = map[Scheme]string{
		HTTP:  "http",
		HTTPS: "https",
	}

	stringToScheme = map[string]Scheme{
		"http":  HTTP,
		"https": HTTPS,
	}
)

type serverConfig struct {
	Host      string `json:"host,omitempty"`
	Protocol  Scheme `json:"protocol,omitempty"`
	HTTPPort  int    `json:"httpPort,omitempty"`
	HTTPSPort int    `json:"httpsPort,omitempty"`
	GRPCPort  int    `json:"grpcPort,omitempty"`
	CertFile  string `json:"certFile,omitempty"`
	CertKey   string `json:"certKey,omitempty"`
}

type databaseConfig struct {
	MigrationsPath string `json:"migrationsPath,omitempty"`
	URL            string `json:"url,omitempty"`
}

func Default() *Config {
	return &Config{
		Log: logConfig{
			Level: "INFO",
		},

		UI: uiConfig{
			Enabled: true,
		},

		Cors: corsConfig{
			Enabled:        false,
			AllowedOrigins: []string{"*"},
		},

		Cache: cacheConfig{
			Memory: memoryCacheConfig{
				Enabled: false,
				Items:   500,
			},
		},

		Server: serverConfig{
			Host:      "0.0.0.0",
			Protocol:  HTTP,
			HTTPPort:  8080,
			HTTPSPort: 443,
			GRPCPort:  9000,
		},

		Database: databaseConfig{
			URL:            "file:/var/opt/flipt/flipt.db",
			MigrationsPath: "/etc/flipt/config/migrations",
		},
	}
}

const (
	// Logging
	cfgLogLevel = "log.level"
	cfgLogFile  = "log.file"

	// UI
	cfgUIEnabled = "ui.enabled"

	// CORS
	cfgCorsEnabled        = "cors.enabled"
	cfgCorsAllowedOrigins = "cors.allowed_origins"

	// Cache
	cfgCacheMemoryEnabled = "cache.memory.enabled"
	cfgCacheMemoryItems   = "cache.memory.items"

	// Server
	cfgServerHost      = "server.host"
	cfgServerProtocol  = "server.protocol"
	cfgServerHTTPPort  = "server.http_port"
	cfgServerHTTPSPort = "server.https_port"
	cfgServerGRPCPort  = "server.grpc_port"
	cfgServerCertFile  = "server.cert_file"
	cfgServerCertKey   = "server.cert_key"

	// DB
	cfgDBURL            = "db.url"
	cfgDBMigrationsPath = "db.migrations.path"
)

func Load(path string) (*Config, error) {
	viper.SetEnvPrefix("FLIPT")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv()

	viper.SetConfigFile(path)

	if err := viper.ReadInConfig(); err != nil {
		return nil, errors.Wrap(err, "loading config")
	}

	cfg := Default()

	// Logging
	if viper.IsSet(cfgLogLevel) {
		cfg.Log.Level = viper.GetString(cfgLogLevel)
	}
	if viper.IsSet(cfgLogFile) {
		cfg.Log.File = viper.GetString(cfgLogFile)
	}

	// UI
	if viper.IsSet(cfgUIEnabled) {
		cfg.UI.Enabled = viper.GetBool(cfgUIEnabled)
	}

	// CORS
	if viper.IsSet(cfgCorsEnabled) {
		cfg.Cors.Enabled = viper.GetBool(cfgCorsEnabled)

		if viper.IsSet(cfgCorsAllowedOrigins) {
			cfg.Cors.AllowedOrigins = viper.GetStringSlice(cfgCorsAllowedOrigins)
		}
	}

	// Cache
	if viper.IsSet(cfgCacheMemoryEnabled) {
		cfg.Cache.Memory.Enabled = viper.GetBool(cfgCacheMemoryEnabled)

		if viper.IsSet(cfgCacheMemoryItems) {
			cfg.Cache.Memory.Items = viper.GetInt(cfgCacheMemoryItems)
		}
	}

	// Server
	if viper.IsSet(cfgServerHost) {
		cfg.Server.Host = viper.GetString(cfgServerHost)
	}
	if viper.IsSet(cfgServerProtocol) {
		cfg.Server.Protocol = stringToScheme[viper.GetString(cfgServerProtocol)]
	}
	if viper.IsSet(cfgServerHTTPPort) {
		cfg.Server.HTTPPort = viper.GetInt(cfgServerHTTPPort)
	}
	if viper.IsSet(cfgServerHTTPSPort) {
		cfg.Server.HTTPSPort = viper.GetInt(cfgServerHTTPSPort)
	}
	if viper.IsSet(cfgServerGRPCPort) {
		cfg.Server.GRPCPort = viper.GetInt(cfgServerGRPCPort)
	}
	if viper.IsSet(cfgServerCertFile) {
		cfg.Server.CertFile = viper.GetString(cfgServerCertFile)
	}
	if viper.IsSet(cfgServerCertKey) {
		cfg.Server.CertKey = viper.GetString(cfgServerCertKey)
	}

	// DB
	if viper.IsSet(cfgDBURL) {
		cfg.Database.URL = viper.GetString(cfgDBURL)
	}
	if viper.IsSet(cfgDBMigrationsPath) {
		cfg.Database.MigrationsPath = viper.GetString(cfgDBMigrationsPath)
	}

	if err := cfg.validate(); err != nil {
		return &Config{}, err
	}

	return cfg, nil
}

func (c *Config) validate() error {
	if c.Server.Protocol == HTTPS {
		if c.Server.CertFile == "" {
			return errors.New("cert_file cannot be empty when using HTTPS")
		}
		if c.Server.CertKey == "" {
			return errors.New("cert_key cannot be empty when using HTTPS")
		}
		if _, err := os.Stat(c.Server.CertFile); os.IsNotExist(err) {
			return fmt.Errorf("cannot find TLS cert_file at %q", c.Server.CertFile)
		}
		if _, err := os.Stat(c.Server.CertKey); os.IsNotExist(err) {
			return fmt.Errorf("cannot find TLS cert_key at %q", c.Server.CertKey)
		}
	}

	return nil
}

func (c *Config) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	out, err := json.Marshal(c)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	if _, err = w.Write(out); err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}
